samples/rust/rust_seq_file.rs

```diff
@@ -0,0 +1,151 @@
+    description: b"Rust sample using a seq_file",
+    license: b"GPL v2",
+}
+
+struct State(Mutex<u32>);
+
+impl seq_file::SeqOperations for State {
+    type Item = String;
+    type Iterator = Take<Repeat<String>>;
+    type DataWrapper = Arc<State>;
+    type IteratorWrapper = Box<Peekable<Self::Iterator>>;
+
+    fn display(item: &Self::Item) -> &str {
+        &item[..]
+    }
+
+    fn start(&self) -> Result<Self::IteratorWrapper> {
+        const MAX_DIGITS: usize = 3;
+        const MAX_LENGTH: usize = MAX_DIGITS + 1;
+        const MAX_COUNT: u32 = 10u32.pow(MAX_DIGITS as u32);
+
+        let count = self.0.lock();
+        let mut message = String::new();
+
+        let template = "rust_seq_file: device opened this many times: ";
+        message.try_reserve_exact(template.len() + MAX_LENGTH)?;
+        // NOPANIC: We reserved space for `template` above.
+        message.push_str(template);
+        if *count < MAX_COUNT {
+            // NOPANIC: There are MAX_LENGTH characters remaining in the string which
+            // leaves space for a MAX_DIGITS digit number and the newline.
+            writeln!(&mut message, "{}", *count).map_err(|_| Error::ENOMEM)?;
+        }
+
+        Box::try_new(repeat(message).take((*count).try_into()?).peekable())
+            .map_err(|_| Error::ENOMEM)
+    }
+}
+
+#[derive(Clone)]
+struct SharedState(Arc<State>);
+
+impl SharedState {
+    fn try_new() -> Result<Self> {
+        let state = Arc::try_new(
+            // SAFETY: `mutex_init!` is called below.
+            unsafe { State(Mutex::new(0)) },
+        )?;
+        // SAFETY: Mutex is pinned behind `Arc`.
+        let pin_state = unsafe { Pin::new_unchecked(&state.0) };
+        mutex_init!(pin_state, "SharedState::0");
+        Ok(SharedState(state))
+    }
+}
+
+impl FileOpener<SharedState> for SharedState {
+    fn open(ctx: &SharedState) -> Result<Self::Wrapper> {
+        pr_info!("rust seq_file was opened!\n");
+        Ok(Box::try_new(ctx.clone())?)
+    }
+}
+
+impl FileOperations for SharedState {
+    type Wrapper = Box<Self>;
+
+    kernel::declare_file_operations!(read);
+
+    fn read<T: IoBufferWriter>(&self, _: &File, data: &mut T, offset: u64) -> Result<usize> {
+        let message = b"incremented read count\n";
+        if offset != 0 {
+            return Ok(0);
+        }
+
+        {
+            let mut count = self.0 .0.lock();
+            *count += 1;
+        }
+
+        data.write_slice(message)?;
+        Ok(message.len())
+    }
+}
+
+struct RustSeqFileDev {
+    _proc: proc_fs::ProcDirEntry<Arc<State>>,
+    _dev: Pin<Box<miscdev::Registration<SharedState>>>,
+}
+
+impl KernelModule for RustSeqFileDev {
+    fn init() -> Result<Self> {
+        pr_info!("Rust seq_file sample (init)\n");
+
+        let state = SharedState::try_new()?;
+
+        let proc_dir_entry = proc_fs::ProcDirEntry::<Arc<State>>::new_seq_private::<State>(
+            c_str!("rust_seq_file"),
+            Arc::clone(&state.0),
```

Comment:
Yeah it does. I've seen it recommend to explicitly use `Arc::clone` or `Rc::clone` when possible just because it emphasizes that the `clone` is just incrementing a reference and not copying allocated data (see e.g. the [docs on `rc`](https://doc.rust-lang.org/stable/std/rc/index.html#cloning-references)). Not sure how prevalent that actually is though so I'm fine switching it to `state.0.clone()`.